### DISCUSSION TITLE
SRP Client Registration with TXT Records

### DISCUSSION THREAD

**Discussion by mmb-davidsmith** [openthread/openthread#8459] [Q&A] [answered]
Hi,
I'm currently trying to build a device which registers itself with an SRP server when it connects to the network. The registration is working, but when I try to add TXT records to my registration, things start to behave in ways that I do not expect.

I start by creating my `otSrpClientBuffersServiceEntry *` variable and initializing it with [otSrpClientBuffersAllocateService](https://openthread.io/reference/group/api-srp#otsrpclientbuffersallocateservice). Reading the documentation, this indicates that it should be configured to point to the single mTxtEntry option.

At this point, I'm looking to add the string `pv=1` to my service entry. When I start to dig into the documentation, it seems like I should have two options to do this.

1. If I set `mKey`, then I should be able to set my value to the string "1" as follows
```
uint16_t size;
uint8_t *txtBuffer = otSrpClientBuffersGetServiceEntryTxtBuffer(serviceEntry, &size);
assert(NULL != txtBuffer);
assert(0 != size);
serviceEntry->mTxtEntry.mKey = "pv";
serviceEntry->mTxtEntry.mValueLength = snprintf((char *)txtBuffer, size, "1");
```
When I do this, the relevant line when calling `srp server service` on my server reads `TXT: [pv=31]`, so it seems like the ascii is getting re-encoded, which the documentation doesn't mention anywhere.

2. Alternatively, the documentation suggests that if I set `mKey` to NULL, then I can encode my string according to RFC6763 section 6.3. I believe the following should also work as a result.
```
uint16_t size;
uint8_t *txtBuffer = otSrpClientBuffersGetServiceEntryTxtBuffer(serviceEntry, &size);
assert(NULL != txtBuffer);
assert(0 != size);
serviceEntry->mTxtEntry.mKey = NULL;
serviceEntry->mTxtEntry.mValueLength = snprintf((char *)txtBuffer, size, "\4pv=1"); // format is <len><string>
```
When I do this, the relevant line when calling `srp server service` on my server reads `TXT: [pv=31]`, so it seems like even "already encoded TXT-DATA" is getting adjusted somewhere.

Am I mis-interpreting the documentation here, or is there a delta between what's documented and what's actually happening which should be fixed?

**Comment by mmb-davidsmith** [comment, marked as answer]
It turns out the TXT data, when decoded is printed as HEX instead of ASCII characters. Would it be worth updating https://github.com/openthread/openthread/blob/main/src/cli/README_SRP_SERVER.md#service to indicate this is the case, particularly because it's printing the key in ASCII but not touching the value.

**Reply by abtink** [reply]
Glad you find it. 

The decoded TXT data value can be any sequence of bytes (including `\0`) so we cannot treat it as a C string therefore we are outputting it as hex bytes.

Here is the method that decodes and outputs the TXT data in CLI:
https://github.com/openthread/openthread/blob/87284f22eb7a03865134de7b02b4425f93241868/src/cli/cli_output.cpp#L229-L272

Yes. Make sense to update the CLI `README_SRP_SERVER` documentation for this. Thanks.

**Reply by jwhui** [reply]
@mmb-davidsmith , would you like to submit a PR updating the `README.md`?

**Reply by mmb-davidsmith** [reply]
@jwhui - I'll try to get something drafted over the next few days.

**Reply by jwhui** [reply]
@mmb-davidsmith , sounds good. Thanks!
